// alliance-redux-api/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "fetchEntities", {
  enumerable: true,
  get: function get() {
    return _actions.fetchEntities;
  }
});
Object.defineProperty(exports, "fetchEntity", {
  enumerable: true,
  get: function get() {
    return _actions.fetchEntity;
  }
});
Object.defineProperty(exports, "addEntity", {
  enumerable: true,
  get: function get() {
    return _actions.addEntity;
  }
});
Object.defineProperty(exports, "addEntityLocal", {
  enumerable: true,
  get: function get() {
    return _actions.addEntityLocal;
  }
});
Object.defineProperty(exports, "updateEntity", {
  enumerable: true,
  get: function get() {
    return _actions.updateEntity;
  }
});
Object.defineProperty(exports, "deleteEntity", {
  enumerable: true,
  get: function get() {
    return _actions.deleteEntity;
  }
});
Object.defineProperty(exports, "deleteEntityLocal", {
  enumerable: true,
  get: function get() {
    return _actions.deleteEntityLocal;
  }
});
Object.defineProperty(exports, "customDetailRoute", {
  enumerable: true,
  get: function get() {
    return _actions.customDetailRoute;
  }
});
Object.defineProperty(exports, "customListRoute", {
  enumerable: true,
  get: function get() {
    return _actions.customListRoute;
  }
});
Object.defineProperty(exports, "updateOptimistic", {
  enumerable: true,
  get: function get() {
    return _actions.updateOptimistic;
  }
});
Object.defineProperty(exports, "Model", {
  enumerable: true,
  get: function get() {
    return _Model["default"];
  }
});
Object.defineProperty(exports, "ForeignKeyField", {
  enumerable: true,
  get: function get() {
    return _ForeignKeyField["default"];
  }
});
Object.defineProperty(exports, "normalize", {
  enumerable: true,
  get: function get() {
    return _normalize["default"];
  }
});
Object.defineProperty(exports, "reduceReducers", {
  enumerable: true,
  get: function get() {
    return _reduceReducers["default"];
  }
});
Object.defineProperty(exports, "initialiseLiveUpdater", {
  enumerable: true,
  get: function get() {
    return _liveUpdate["default"];
  }
});
Object.defineProperty(exports, "Field", {
  enumerable: true,
  get: function get() {
    return _Field["default"];
  }
});
Object.defineProperty(exports, "createReducer", {
  enumerable: true,
  get: function get() {
    return _createReducer["default"];
  }
});
Object.defineProperty(exports, "createModelReducer", {
  enumerable: true,
  get: function get() {
    return _createModelReducer["default"];
  }
});
Object.defineProperty(exports, "createModelsReducer", {
  enumerable: true,
  get: function get() {
    return _createModelsReducer["default"];
  }
});
exports.api = void 0;

var _api = require("./api");

var _actions = require("./actions");

var _Model = _interopRequireDefault(require("./Model"));

var _ForeignKeyField = _interopRequireDefault(require("./ForeignKeyField"));

var _normalize = _interopRequireDefault(require("./normalize"));

var _reduceReducers = _interopRequireDefault(require("./reduceReducers"));

var _liveUpdate = _interopRequireDefault(require("./liveUpdate"));

var _Field = _interopRequireDefault(require("./Field"));

var _createReducer = _interopRequireDefault(require("./createReducer"));

var _createModelReducer = _interopRequireDefault(require("./createModelReducer"));

var _createModelsReducer = _interopRequireDefault(require("./createModelsReducer"));

var api = {
  post: _api.post,
  get: _api.get,
  del: _api.del,
  patch: _api.patch,
  callApi: _api.callApi,
  API_CALL_AUTH_REQUIRED: _api.API_CALL_AUTH_REQUIRED
};
exports.api = api;
