// alliance-redux-api/.eslintrc.js
const assert = require('assert');
const path = require('path');
const restrictedGlobals = require('confusing-browser-globals');
// In dev it's annoying to have these lint errors shown - it's really only in CI we care about
// it given the reformat commit hook should fix this anyway
const lintPrettier = process.env.LINT_PRETTIER;
const globals = {
    global: true,
    __DEBUG__: false,
    __DEBUG_NEW_WINDOW__: false,
    spyOn: true,
};
const rules = {
    'no-restricted-globals': ['error'].concat(restrictedGlobals),
    'flowtype/space-after-type-colon': 0,
    'import/no-extraneous-dependencies': 0,
};
if (lintPrettier) {
    rules['prettier/prettier'] = 'error';
}
module.exports = {
    parser: 'babel-eslint',
    extends: [
        '@alliance-software/eslint-config',
        'plugin:flowtype/recommended',
        'prettier',
        'prettier/flowtype',
    ],
    plugins: ['flowtype', lintPrettier && 'prettier'].filter(Boolean),
    rules,
    globals,
    env: {
        browser: true,
        jest: true,
    },
};
