// @alliance-software/djrad/model/fields/DateRangeField.js
// @flow
import React from 'react';
import moment from 'moment';
import DateFormatter from '../../components/formatter/DateFormatter';
import RangeField from './RangeField';
import type { WidgetProps } from '../../components/form/FormField';
import type { FormatterProps } from '../../components/formatter/types';
import DateRangeWidget from '../../components/form/widgets/DateRangeWidget';

export default class DateRangeField extends RangeField {
    parse(value: any, name: string) {
        if (name === this.name) {
            const { lower, upper } = value;
            return {
                upper: moment.isMoment(upper) ? upper.format('YYYY-MM-DD') : upper,
                lower: moment.isMoment(lower) ? lower.format('YYYY-MM-DD') : lower,
            };
        }
        return moment.isMoment(value) ? value.format('YYYY-MM-DD') : value;
    }

    format(value: any, name: string) {
        if (!value) {
            return null;
        }
        // Dealing with the object ie. { lower: '...', upper: '...' }
        if (name === this.name) {
            return {
                lower: value.lower ? moment(value.lower) : null,
                upper: value.upper ? moment(value.upper) : null,
            };
        }
        // Otherwise dealing with the individual field (lower or upper)
        return moment(value);
    }

    getRangeValueFormatter(props: FormatterProps) {
        return <DateFormatter {...props} />;
    }

    getDefaultWidget(props: WidgetProps) {
        return <DateRangeWidget {...props} />;
    }
}
