// @alliance-software/djrad/components/form/widgets/InputRangeWidget.js
import * as React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import { Input } from 'antd';
import componentPropType from '../../../prop-types/component';
import { fieldShape } from '../../../prop-types/form';

import './InputRangeWidget.less';

export default function InputRangeWidget(props) {
    const {
        inputWidget: InputWidget,
        inputWidgetProps,
        upper,
        lower,
        className,
        lowerPlaceholder,
        upperPlaceholder,
        inputClassName,
        separatorText,
    } = props;
    return (
        <Input.Group compact className={cx('djrad--range-widget', className)}>
            <InputWidget
                {...lower.input}
                {...inputWidgetProps}
                placeholder={lowerPlaceholder}
                className={cx('djrad--range-widget__lower', inputClassName)}
            />
            <Input
                placeholder={separatorText}
                disabled
                className={cx('djrad--range-widget__separator', inputClassName)}
            />
            <InputWidget
                {...upper.input}
                {...inputWidgetProps}
                placeholder={upperPlaceholder}
                className={cx('djrad--range-widget__upper', inputClassName)}
            />
        </Input.Group>
    );
}

InputRangeWidget.propTypes = {
    /** Class to pass to lower & upper bound widgets */
    inputClassName: PropTypes.string,
    /** Component to use for the lower and upper bound input widgets */
    inputWidget: componentPropType,
    /** Extra props to pass through to inputWidget */
    inputWidgetProps: PropTypes.object,
    /** Text to use as a separator between the lower and upper bounds. */
    separatorText: PropTypes.string,
    /** Placeholder text to use for lower bound */
    lowerPlaceholder: PropTypes.string,
    /** Placeholder text to use for upper bound */
    upperPlaceholder: PropTypes.string,
    /** Lower bound field. Typically comes from redux-form, see https://redux-form.com/7.3.0/docs/api/field.md/#props */
    lower: fieldShape.isRequired,
    /** Upper bound field. Typically comes from redux-form, see https://redux-form.com/7.3.0/docs/api/field.md/#props */
    upper: fieldShape.isRequired,
    /** Class name for group wrapper */
    className: PropTypes.string,
};

InputRangeWidget.defaultProps = {
    inputWidget: Input,
    separatorText: '~',
};
