// @alliance-software/djrad/api/types.js
// @flow
import type { List } from 'immutablejs'; // eslint-disable-line
import type { Model } from '../model/Model';

export type PaginatedListResponsePayload = {
    // Total number of results (not count for page)
    count: number,
    // The URL to use for next page
    next: string,
    // The URL to use for previous page
    previous: string,
    // Array of individual objects for response
    results: Array<{}>,
};

export type UnpaginatedListResponsePayload = Array<{}>;

export type ApiListResponse = {
    // Optional meta object that was passed with original call
    meta: ?{},
    payload: PaginatedListResponsePayload | UnpaginatedListResponsePayload,
    // The redux action type
    type: string,
};

export type ModelApiListResponse = {
    count: number,
    // The URL to use for next page
    next: string,
    nextPage: ?number,
    // The URL to use for previous page
    previous: string,
    previousPage: ?number,
    // Array of raw objects from response. Entities below contains same data but as Model instances.
    results: Array<{}>,
    // The ids of each record in results.
    ids: Array<number | string>,
    // List of model record instances
    entities: List<Model>,
};
