// @alliance-software/djrad-devtools/SourceLink.js
import React, { useCallback } from 'react';

// Comes from here
// import launchEditorEndpoint from 'react-dev-utils/launchEditorEndpoint';
const launchEditorEndpoint = '/__open-stack-frame-in-editor';

const styles = {
    cursor: 'pointer',
    border: 0,
    padding: 0,
    margin: 0,
    color: '#1890ff',
};

export default function SourceLink({ source, fixedWidth, children, style }) {
    const { fileName, filename = fileName, lineNumber, colNumber } = source;
    const handleClick = useCallback(
        () =>
            fetch(
                // eslint-disable-next-line
                process.env.PUBLIC_URL.replace(/\/$/, '') +
                    launchEditorEndpoint +
                    '?fileName=' +
                    window.encodeURIComponent(filename.replace('webpack:///', '')) +
                    '&lineNumber=' +
                    window.encodeURIComponent(lineNumber || 1) +
                    '&colNumber=' +
                    window.encodeURIComponent(colNumber || 1)
            ),
        [colNumber, filename, lineNumber]
    );
    const finalStyles = { ...styles, ...style };
    if (fixedWidth) {
        Object.assign(finalStyles, {
            width: fixedWidth,
            display: 'inline-block',
            textOverflow: 'ellipsis',
            overflow: 'hidden',
            whiteSpace: 'nowrap',
        });
    }
    return (
        <button style={finalStyles} onClick={handleClick}>
            {children || `${filename}:${lineNumber}`}
        </button>
    );
}
